Allow setting withCredentials in spf xhr options

zhaoz opened the pull request:
This allows SPF to be used with CORS + Cookies.

nicksay commented:
You're going to have to thread this to `spf.nav.request.send` as well.  To make this available to clients, we could also thread this one step further to `spf.nav.load`.  (Since `withCredentials` has no-effect on same-domain requests and we currently force reloads on cross-origin navigation, I don't think we need to make this available to `spf.nav.navigate`.)

zhaoz commented:
Added commits to forward withCredentials to spf.nav.request.send and spf.nav.load, also added to the markdown file.

Let me know if you prefer to have this squashed into one commit.

nicksay commented on doc/api.md:
```diff
@@ -219,6 +219,8 @@ to the callback will be an object that conforms to the
 `postData: ArrayBuffer | Blob | Document | FormData | null | string | undefined`  
 Optional data to send with the request.  Only used if the method is "POST".  
+`withCredentials: boolean | undefined`
+Optional flag to configure the XHR to send withCredentials or not.
```
You should revert this file; it will be auto-generated from src/api.js with the next release and will use the description from the JsDoc.

nicksay commented on src/api.js:
```diff
@@ -326,6 +326,13 @@ spf.RequestOptions.prototype.postData;
 
 
 /**
+ * If true, will set withCredentials on XHR.
```
Update to the version you had in `api.md`?

nicksay commented:
Please do squash into a single commit.  You either do it beforehand in your fork or "squash and merge" using the GitHub UI when merging the PR.

zhaoz commented:
Squashed my commits and addressed comments. Not sure how to mark comments as addressed though.

nicksay commented:
Ah, sorry I forgot this before -- can you add the `withCredentials` property to the typedef in base.js too?  https://github.com/youtube/spfjs/blob/master/src/client/base.js#L301

zhaoz commented:
Updated

rviscomi commented on src/client/nav/request.js:
```diff
@@ -65,7 +65,8 @@ goog.require('spf.url');
  *   referer: (string|null|undefined),
- *   type: (string|undefined)
+ *   type: (string|undefined),
+ *   withCredentials: (boolean|undefined)
```
Add a corresponding description to the comment above (line ~55).

zhaoz replied:
Done.

rviscomi commented on src/client/net/xhr.js:
```diff
@@ -38,7 +38,8 @@ goog.require('spf');
  *   responseType: (string|undefined),
- *   timeoutMs: (number|undefined)
+ *   timeoutMs: (number|undefined),
+ *   withCredentials: (boolean|undefined)
```
Same (line ~33)

zhaoz replied:
Done.

rviscomi commented on src/client/base.js:
```diff
@@ -295,7 +295,8 @@ spf.MultipartResponse;
  *   onRequest: (function(spf.EventDetail)|undefined),
- *   postData: (ArrayBuffer|Blob|Document|FormData|null|string|undefined)
+ *   postData: (ArrayBuffer|Blob|Document|FormData|null|string|undefined),
+ *   withCredentials: (boolean|undefined)
```
Same (line ~288)

zhaoz replied:
Done.

rviscomi commented:
LGTM after nits.

zhaoz commented:
Addressed comment issue. And updated.
